File: wysepassword.js

/* 
    Javascript code to encode or decode a Dell Wyse password
    Reverse engineering by David Lodge

    Converted to Javascript by Ryan Maloney

    Usage: var EncodedPassword = NFuseEncode("password")
           var DecodedPassword = NFuseDecode("NFBBMHBBMDAJNOBP") 
*/
// Encode function
// Input: String
// Returns: String
function NFuseEncode(Value){
    var Length = Value.length;
    var arrReturn = new Array(Length);
    var arrTemp = new Array(Length);
    for (let i=0; i<Length; i++) {
        let a,b,y
        a = Value.charAt(i).charCodeAt();
        a ^= 0xA5;
        a ^= arrTemp[i-1];
        arrTemp[i] = a;
        y = i*2;
        b = a;
        b &= 0x0F;
        b += 0x41;
        a >>= 4;
        a += 0x41;
        arrReturn[y] = String.fromCharCode(a);
        arrReturn[y+1] = String.fromCharCode(b);
    }
    return arrReturn.join("");
}
// Decode function
// Input: String
// Returns: String
function NFuseDecode(Value){
    var Length = Value.length/2;
    var arrReturn = new Array(Length);
    for (let i=0; i<Length; i++) {
        let a = Value.charAt(i*2).charCodeAt();
        a -= 1;
        a <<= 4;
        a += Value.charAt(i*2+1).charCodeAt();
        a -= 0x41;
        arrReturn[i] = a;  
    }
    for (let i=Length-1; i>=0; i--) {
        let a = (arrReturn[i-1] === undefined)?1024:arrReturn[i-1];
        a ^= arrReturn[i];
        a ^= 0xA5;
        arrReturn[i] = String.fromCharCode(a);
    }
    return arrReturn.join("");
}

File: wysepassword.c
// Basic C code to decrypt Wyse password for decoding
// Reverse engineering and C mangulation by David Lodge
// 
// Extended to include an encoder by Ryan Maloney

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

void NFuseDecode(char *src, int length, unsigned char *dst)
{
   unsigned char a;
   int i;

   length/=2;

   for (i=0; i<length; i++)
   {
      a=src[i*2];
      a-=1;
      a<<=4;
      a+=src[i*2+1];
      a-=0x41;
      dst[i]=a;
   }

   for (i=length; i>=0; i--)
   {
      a=dst[i-1];
      a^=dst[i];
      a^=0xA5;
      dst[i]=a;
   }
   dst[length]='\0';
}

void NFuseEncode(char *src, int length, unsigned char *dst)
{
  unsigned char a,b,temp[256];
  int i,y;

  for (i=0; i<length; i++)
  {
    a=src[i];
    a^=0xA5;
    a^=temp[i-1];
    temp[i]=a;
    y=i*2;
    b=a;
    b=b & 0x0F;
    b+=0x41;
    a>>=4;
    a+=0x41; 
    dst[y]=a;
    dst[y+1]=b;
  }
   dst[length*2]='\0';
}


int main(int argc, char **argv)
{
   unsigned char buffer[256];
   int opt;
   
   if (argc < 2) {
	   printf("Usage: %s [OPTION] PASSWORD\nExample: %s -e somepassword\
			   \nOptions:\n -d\t\tEncode password\n -e\t\tDecode Password\n\n", argv[0],argv[0]);
	   return -1;
   }
  
   while ((opt = getopt(argc, argv, "e:d:")) != -1) {
	switch(opt) {
		case 'e':
   			NFuseEncode(optarg,strlen(optarg), buffer);
   			printf("%s\n", buffer);
			break;
		case 'd':
			NFuseDecode(optarg,strlen(optarg), buffer);
   			printf("%s\n", buffer);
			break;
		case ':':
			printf("Option needs a value\n");
			break;
	}
   }
   return 0;
}
